fixes issues with `v`

netbonus opened the pull request.

netbonus committed:
fixes issues with `v`

netbonus committed:
fix lint

yilmazbahadir commented on src/genericSigning.ts:
```diff
@@ -232,21 +232,23 @@ export const parseGenericSigningResponse = function (res, off, req) {
+      req.hashType === Constants.SIGNING.HASHES.KECCAK256
+    ) {
+      const vBn = getV(req.origPayloadBuf, parsed);
+      parsed.sig.v = BigInt(vBn.toString());
```
I see the `viem/Signature` implementation and also the legacy eth transactions have the type of `v` as `bigint` :white_check_mark:
Should we make it optional and update other places to `bigint` like https://github.com/GridPlus/gridplus-sdk/blob/6790828e26b8a9425a5a915c7329b64831d80497/docs/docs/signing.md?plain=1#L203 ?

And also where do we have the following definition, that the Kalo mentioned in his [issue](https://github.com/GridPlus/gridplus-sdk/issues/611)?

```interface SignData {
  sig?: {
    v: Buffer;
    r: Buffer;
    s: Buffer;
  };
  ...
}
```

yilmazbahadir replied:
WDYT about this one @netbonus ?

yilmazbahadir commented:
@netbonus please see Kalo's test result below:

lib sign method fails (throws):

Error: Could not recover V. Bad transaction data.
    at getV (util.ts:749:1)
    at parseGenericSigningResponse (genericSigning.ts:250:1)
    at decodeSignResponse (sign.ts:300:1)
    at sign (sign.ts:86:1)
This happens right after I approve the plain text message on the device (all visualized correctly).

My req is:

{
    "signerPath": [2147483692, 2147483708, 2147483648, 0, 0],
    "payload": "0x5465737421",
    "curveType": 0,
    "hashType": 1
}
(Attempting to sign a plain text message using general signing).

superKalo commented:
@netbonus additional details: I'm on the latest firmware. Also when I trigger `sign` with the req that @yilmazbahadir shared, the plain message appears on the Lattice device (all visualized correctly), I approve it and then - it throws.

netbonus committed:
fix: improve v parameter handling and transaction validation

- Add comprehensive transaction schema validation using Zod
- Fix v parameter calculation for different transaction types
- Consolidate signature recovery logic into unified functions
- Add transaction parsing utilities for better error handling
- Improve test coverage for signature utils and validators
- Fix data field encoding in test builders
- Update encoder test snapshots to reflect corrected behavior

This addresses issues with v parameter calculation across different
transaction types (legacy, EIP-155, EIP-1559, EIP-2930, EIP-7702)
and improves the overall robustness of transaction handling.

netbonus committed:
fix lock

yilmazbahadir commented on src/__test__/unit/parseGenericSigningResponse.test.ts:
```diff
@@ -0,0 +1,173 @@
+
+  it('should handle RLP-encoded data that looks like a transaction', () => {
+    // Create an RLP-encoded array with 6+ elements (looks like a transaction)
+    const RLP = require('@ethereumjs/rlp').RLP;
```
Should we use a static import instead?

yilmazbahadir commented on src/__test__/unit/parseGenericSigningResponse.test.ts:
```diff
@@ -0,0 +1,173 @@
+
+    // For non-EVM generic messages, v should be 27 or 28
+    const vNumber = Number(result.sig.v);
+    expect([27n, 28n]).toContain(result.sig.v);
```
```suggestion
    expect([27, 28]).toContain(vNumber);
```

yilmazbahadir commented on src/__test__/unit/signatureUtils.test.ts:
```diff
@@ -365,3 +365,87 @@ describe('getYParity', () => {
+      pubkey: Buffer.from('04' + '1'.repeat(128), 'hex'),
+    };
+
+    // Should throw because pubkey doesn't match
```
Should we use a more explanatory description? `should parse viem parseTransaction correctly` seems not to be what we test here.

yilmazbahadir commented on src/__test__/unit/signatureUtils.test.ts:
```diff
@@ -365,3 +365,87 @@ describe('getYParity', () => {
+    };
+
+    // Should throw because signature doesn't match
+    expect(() => getV(txHex, mockResp)).toThrow();
```
same? (we should either update the test case description or the assertion here)

yilmazbahadir commented on src/util.ts:
```diff
@@ -708,86 +708,177 @@ export const generateAppSecret = (
+export const convertRecoveryToV = function (
+  recovery: number,
+  txData: any = {},
+) {
```
should we add a return type here, so that the consumers can always know to check the result type.

netbonus committed:
address feedback

yilmazbahadir commented on src/__test__/unit/parseGenericSigningResponse.test.ts:
```diff
@@ -0,0 +1,172 @@
+    expect(typeof result.sig.v).toBe('bigint');
+
+    // For non-EVM generic messages, v should be 27 or 28
+    expect([27, 28]).toContain(result.sig.v);
```
The test is failing, I reckon we need the following change:

```suggestion
    const vNumber = Number(result.sig.v);
    expect([27, 28]).toContain(vNumber);
```

netbonus committed:
fix bigint
